### PR TITLE
Allow configuration MaxGracefulTerminationSec flag on ClusterAutoscaler

### LINKED ISSUE #4695: Keep CA max-graceful-termination-sec configurable
**How to categorize this issue?**

/area auto-scaling
/kind enhancement
/priority 3

**What would you like to be added**:
Keep the [max-graceful-termination-sec](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#does-ca-respect-gracefultermination-in-scale-down) configurable at the Gardener. 

**Why is this needed**:
During scale down of nodes in the cluster, the cluster autoscaler waits for a maximum of 10 minutes for any pod(s) to wait for graceful termination. However, this is a [configurable flag at the cluster autoscaler](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#does-ca-respect-gracefultermination-in-scale-down) to let the shoot owners decide the maximum time to wait while draining the node. The eventual plan would be to delegate the task of draining [completely to the cluster autoscaler](https://github.com/gardener/autoscaler/issues/99). In the meanwhile, this enhancement would address this.  

/assign
/status in-progress

### PR DESCRIPTION
**Pull request by prashanth26**
**How to categorize this PR?**

/area auto-scaling
/kind enhancement

**What this PR does / why we need it**:
During scale down of nodes in the cluster, the cluster autoscaler waits for a maximum of 10 minutes for any pod(s) to wait for graceful termination. However, this is a [configurable flag at the cluster autoscaler](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#does-ca-respect-gracefultermination-in-scale-down) to let the shoot owners decide the maximum time to wait while draining the node.  

**Which issue(s) this PR fixes**:
Fixes #4695 

**Special notes for your reviewer**:
The eventual plan would be to delegate the task of draining [completely to the cluster autoscaler](https://github.com/gardener/autoscaler/issues/99). In the meanwhile, this enhancement would give some breathing space for end-users to have a handle. 

**Release note**:

```feature user
Allows configuration of `MaxGracefulTerminationSeconds` flag on ClusterAutoscaler. This allows end-users to configure maximum graceful termination (drain) seconds beyond which the node is force deleted during scale-down of cluster nodes. The default value is 600 seconds. 
```

### CODE REVIEW

File: pkg/apis/core/types_shoot.go [hunk 1]
```diff
@@ -333,6 +333,8 @@ type ClusterAutoscaler struct {
 	Expander *ExpanderMode
 	// MaxNodeProvisionTime defines how long CA waits for node to be provisioned (default: 20 mins).
 	MaxNodeProvisionTime *metav1.Duration
+	// MaxGracefulTerminationSec number of seconds CA waits for pod termination when trying to scale down a node (default: 600).
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]:
```suggestion
	// MaxGracefulTerminationSeconds is the number of seconds CA waits for pod termination when trying to scale down a node (default: 600).
```

File: example/90-shoot.yaml [hunk 2]
```diff
@@ -232,6 +232,7 @@ spec:
   #     SomeKubernetesFeature: true
   # clusterAutoscaler:
   #   expander: "least-waste" # see: https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#what-are-expanders
+  #   maxGracefulTerminationSec: 600
```
Comment (rfranzke) [inline example/90-shoot.yaml]:
```suggestion
  #   maxGracefulTerminationSeconds: 600
```

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 3]
```diff
@@ -396,6 +396,9 @@ type ClusterAutoscaler struct {
 	// MaxNodeProvisionTime defines how long CA waits for node to be provisioned (default: 20 mins).
 	// +optional
 	MaxNodeProvisionTime *metav1.Duration `json:"maxNodeProvisionTime,omitempty" protobuf:"bytes,8,opt,name=maxNodeProvisionTime"`
+	// MaxGracefulTerminationSec number of seconds CA waits for pod termination when trying to scale down a node (default: 600).
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/types_shoot.go]:
```suggestion
	// MaxGracefulTerminationSeconds is the number of seconds CA waits for pod termination when trying to scale down a node (default: 600).
```

File: pkg/apis/core/types_shoot.go [hunk 4]
```diff
@@ -333,6 +333,8 @@ type ClusterAutoscaler struct {
...
+	MaxGracefulTerminationSec *int32
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]:
```suggestion
	MaxGracefulTerminationSeconds *int32
```

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 5]
```diff
@@ -396,6 +396,9 @@ type ClusterAutoscaler struct {
...
+	// +optional
+	MaxGracefulTerminationSec *int32 `json:"maxGracefulTerminationSec,omitempty" protobuf:"varint,9,opt,name=maxGracefulTerminationSec"`
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/types_shoot.go]:
```suggestion
	MaxGracefulTerminationSeconds *int32 `json:"maxGracefulTerminationSeconds,omitempty" protobuf:"varint,9,opt,name=maxGracefulTerminationSeconds"`
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 6]
```diff
@@ -408,6 +408,9 @@ type ClusterAutoscaler struct {
 	// MaxNodeProvisionTime defines how long CA waits for node to be provisioned (default: 20 mins).
 	// +optional
 	MaxNodeProvisionTime *metav1.Duration `json:"maxNodeProvisionTime,omitempty" protobuf:"bytes,8,opt,name=maxNodeProvisionTime"`
+	// MaxGracefulTerminationSec number of seconds CA waits for pod termination when trying to scale down a node (default: 600).
+	// +optional
+	MaxGracefulTerminationSec *int32 `json:"maxGracefulTerminationSec,omitempty" protobuf:"varint,9,opt,name=maxGracefulTerminationSec"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// MaxGracefulTerminationSeconds is the number of seconds CA waits for pod termination when trying to scale down a node (default: 600).
	// +optional
	MaxGracefulTerminationSeconds *int32 `json:"maxGracefulTerminationSeconds,omitempty" protobuf:"varint,9,opt,name=maxGracefulTerminationSeconds"`
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/apis/core/types_shoot.go (hunk 1 above)
Comment (prashanth26) [reply pkg/apis/core/types_shoot.go]: I was also wondering about this naming convention. Kept it this way as it matches with the CA flag name. However, I agree that `MaxGracefulTerminationSeconds` is easier to understand. I will make this change.

Comment (rfranzke) [reply pkg/apis/core/types_shoot.go]: Thanks!

Comment (prashanth26) [reply pkg/apis/core/types_shoot.go]: Made both suggested changes of renaming var and updating doc string with [this commit](https://github.com/gardener/gardener/pull/4697/commits/a849f80f77b6b70c2f7fc080c53122926e70b412)

File: pkg/apis/core/validation/shoot_test.go [hunk 7]
```diff
@@ -1842,11 +1842,13 @@ var _ = Describe("Shoot Validation Tests", func() {
...
 
 		var (
-			negativeDuration   = metav1.Duration{Duration: -time.Second}
-			expanderLeastWaste = core.ClusterAutoscalerExpanderLeastWaste
-			expanderMostPods   = core.ClusterAutoscalerExpanderMostPods
-			expanderPriority   = core.ClusterAutoscalerExpanderPriority
-			expanderRandom     = core.ClusterAutoscalerExpanderRandom
+			negativeDuration         = metav1.Duration{Duration: -time.Second}
+			negativeInteger    int32 = -100
+			postiveInteger     int32 = 100
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot_test.go]:
```suggestion
			positiveInteger     int32 = 100
```

**Review by rfranzke — DISMISSED** [review]
/lgtm
@prashanth26 do you plan to tackle @ialidzhikov's suggestion?

**Comment by prashanth26** [comment]
> /lgtm
> @prashanth26 do you plan to tackle @ialidzhikov's suggestion? 

Apologies. I somehow forgot about the PR. Took in suggested changes. 
/squash

**Comment by rfranzke** [comment]
@prashanth26 Can you  check why the failing `verify` step is   a flake or related to this PR?

**Comment by prashanth26** [comment]
> @prashanth26 Can you check why the failing `verify` step is a flake or related to this PR?

Will check it out.

**Comment by prashanth26** [comment]
> @prashanth26 Can you check why the failing `verify` step is a flake or related to this PR?

One of the test variables wasn't renamed. I made the change and now the pipeline passes. PTAL. Also squashed all the changes.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**